### ISSUE TITLE
State set via useState does not update in custom validation function.

### ISSUE THREAD

**Issue by nobleach** [react-hook-form/react-hook-form#342]
**Describe the bug**
When registering a custom validator that uses local component state, only the initial values are respected when that function is called. Validation mode is set to 'onBlur'. Bluring the field does cause the custom validator to run, but the values do not reflect the actual component state.

**To Reproduce**
See this codesandbox: https://codesandbox.io/s/stoic-sun-lm3sg with the console open.

**Expected behavior**
Values in the custom validator should update every time the local state is updated.

**Desktop (please complete the following information):**
 - OS: Linux
 - Browser Chrome 77

**Additional context**

**Comment by bluebill1049** [comment]
hey @nobleach 

just took a quick look on this example, curious on this part
```
const validatePassword = () => {
    const allCriteriaMet = [
      hasCapitalLetter,
      hasLowerCaseLetter,
      hasNumber,
      hasMinimumChars
    ].every(v => v === true);

    // none of these are updating on each render
    console.log("---Inside validate function----");
    console.log("hasMinimumChars", hasMinimumChars);
    console.log("hasNumber", hasNumber);
    console.log("hasLowerCaseLetter", hasLowerCaseLetter);
    console.log("hasCapitalLetter", hasCapitalLetter);

    if (allCriteriaMet) {
      return;
    }
  };```

why you are not using the value from the validate function>

https://codesandbox.io/s/8n937m64o9

**Comment by nobleach** [comment]
In the real implementation, I'm also checking a "confirm password" input. I'm checking that the above criteria are true as the user types. I call the function in question `onBlur`. I certainly can re-work this to follow best practices however. Oddly though, when I remove react-hook-form, and just use plain React hooks with `onChange` and `onUpdate`, the state updates correctly. Could this be a problem with the ref that's created?

**Comment by bluebill1049** [comment]
I will take a close look when I am on my computer :)👍🏻

**Comment by nobleach** [comment]
Thank you!

**Comment by bluebill1049** [comment]
This works for me: https://codesandbox.io/s/elegant-einstein-gvc7u

btw you set the mode to be `onBlur` so validation is trigger on `onBlur`

**Comment by nobleach** [comment]
Yes, the `onChange` was only for updating things like `Has at least 1 capital letter`, `Has at least 1 number` as the user types. If all those criteria are met, when I leave the component (probably to click the submit button) react-hook-form's validation should take over and either allow or disallow submitting the form. I see what you're doing by using the value inside the `validatePassword` function, and it basically mirrors what's going on in `checkRuleSet`...  I'll rework my code. Thanks for taking a look, and my apologies for submitting a bug ticket when this may not be an actual bug.

**Comment by bluebill1049** [comment]
not at all @nobleach for your use case, I used to have a PR for support multiple errors

https://github.com/react-hook-form/react-hook-form/pull/311

but got closed, it probably would be useful for multiple errors returns like your use case

**Comment by bluebill1049** [comment]
maybe I should relook at that PR and see if it's useful to others

**Comment by bluebill1049** [comment]
BTW, i am still looking at your example and try to work out a good solution

**Comment by nobleach** [comment]
While I certainly appreciate you looking, don't feel like you have to solve my problem. I was just concerned that the problem had to do with registering a ref. 

This example works when not registering a ref via `register()`. https://codesandbox.io/s/stoic-sun-lm3sg (it's not using react-hook-form at all though)

**Comment by bluebill1049** [comment]
hahaha it's my pleasure to help you out here, because someone else may have the same proble. so no worries at all :)

hey, i have this idea:
```
const validatePassword = value => {
    let noError = true;
    if (value.length >= 1) {
      setError("xxx", "something is wrong");
      noError = false;
    } else {
      clearError("xxx");
    }

    return noError;
  };```

**Comment by nobleach** [comment]
Yeah, I'll play with that this evening. Thanks again for all the help.

**Comment by bluebill1049** [comment]
no worries :) if you still having issues, you can ask in here too https://spectrum.chat/react-hook-form

**Comment by nobleach** [comment]
Perfect!

**Comment by dteiml** [comment]
So I've come to this issue by searching for `useState` in closed issues.
I'm not sure if it's expected behavior, but I have the same problem as @nobleach . Validate functions seem to run at mount time - closure-wise. You can parameterize them with the input, but it seems to me that nothing else.

This generalizes @nobleach 's observation. `useState` won't work, but so won't something like this:

```
register: ({validate: (text) => (!loading) && (R.length (results) === 1) && (text === results[0].text) || `problem`})
```

(My use case is an asynchronous dropdown. ... I want the validation to succeed if the input value is the only possibility in the dropdown. But since dropdown comes from API, I can't figure out a way to do that with register. Gonna refactor to `yup` now)

**Comment by bluebill1049** [comment]
@dteiml there are three ways to fix this, 
1. you can store your state value in useRef
2. try to use watch() to store the value in a input and call watch during validate function
3. hoist a variable outside the react component 
eg.
```
let data = ''
function test() {
  data = xxx
}
```

**Comment by khamphou** [comment]
I was facing same issue and behavior as you, and I have sorted it out using the 'bind' trick below inside an async rule check.
spoRegion is a destructured state property, that triggers the ValidatorForm.addValidationRule erach time it is modified.
the spoRegion value got updated each time inside the rule check, following the update of another input field which has a dependency with the spoRegion value.

Hope it helps.

useEffect(() => {
        ValidatorForm.bind(spoRegion)
        ValidatorForm.addValidationRule('urlExists', (value) => {
            return (async () => {
                const checkResult = await checkSite(value, spoRegion)
                    return !checkResult.exists
            })()
        })
    }, [state.spoRegion])


Now, i am facing another problem, as I need to validate the input url value, each time I changed the spoRegion value.
The validation rule is not triggered as the input value is not modified, any help or code would be appreciated to call the validation of this input field.
